src/inmanta/data/__init__.py

```diff
@@ -2298,13 +2636,158 @@ async def get_resources_for_version_raw(
         return resources
 
     @classmethod
-    async def get_latest_version(cls, environment: uuid.UUID, resource_id: ResourceIdStr) -> Optional["Resource"]:
+    async def get_latest_version(cls, environment: uuid.UUID, resource_id: m.ResourceIdStr) -> Optional["Resource"]:
         resources = await cls.get_list(
             order_by_column="model", order="DESC", limit=1, environment=environment, resource_id=resource_id
         )
         if len(resources) > 0:
             return resources[0]
 
+    @classmethod
+    def _get_released_resources_partial_base_query(cls) -> str:
+        """A partial query describing the conditions for selecting the latest released resources,
+        according to the model version number."""
+        return f"""
+            FROM (
+                SELECT DISTINCT ON (r.resource_id) r.resource_id, r.attributes, r.resource_version_id, r.resource_type,
+                    r.agent, r.resource_id_value, r.environment,
+                    (CASE WHEN
+                            (SELECT r.model < MAX(public.configurationmodel.version)
+                            FROM public.configurationmodel
+                            WHERE public.configurationmodel.released=TRUE
+                            AND environment = $1)
```

Comment:
This query is very tricky, with respect to the `$1`. This query works because the caller of the `get_released_resources()` method happens to insert the constraint on the environment first into the query dictionary and then the other constraints. Since, an iteration over a dictionary is consistent with the order of insertion, $1 is correct here. But I think it's a bad idea to rely on this for the following reasons:

* The code is very hard to read.
* The fact that the order of iteration over a dictionary is consistent with the order if insertion, is a implementation detail in python3.6. The documentation explicitly mentions that this behavior may vary across Python implementation. Only from Python3.7, this is an officially supported feature.

Reply:
I'll try to improve the situation, you're right